### PR TITLE
ASoC: SOF/Intel/AMD/Mediatek/IMX: enable GCC '-fanalyzer' checks

### PR DESCRIPTION
**Pull request by plbossart**
Cleanups to work-around issues, prior to enable this static analysis by default.

It's pretty interesting that none of the other tools reported these issues....

@dbaluta @yaochunhung @kuanhsuncheng @crojewsk-intel @ujfalusi @ranj063 @marc-hb feedback welcome.

### CODE REVIEW

**Comment by marc-hb** [comment]
https://github.com/thesofproject/linux/actions/runs/4897771652/jobs/8746195783?pr=4340 failed like this:
```
gcc: error: unrecognized command line option ‘-fanalyzer’
```

Try an Ubuntu upgrade like this one:
- https://github.com/thesofproject/sof/pull/7563

Even if not enough it won't hurt anyway!

**Comment by marc-hb** [comment]
> It's not a problem today, since we do
check in the first use of the function, but static analysis tools are
not aware of the different ALSA stages. Rather than argue forever,
let's just add the error checks consistently and make sure this tool
can be added to the CI checks.

I like this pragmatic approach. I would add that it's NOT just to please static analyzers and silence false positives:

- This also saves time for human code reviewers / "human analyzers"
- Defensive programming makes the code more future proof. The conditions that are true today may be different tomorrow.


> This is a false-positive, the GCC analyzer generated that report by
considering if (bytes->block) as true in some cases and false in
others.

This one looks more like a relatively simple bug that should be reported to GCC? Does it still show with cutting edge GCC? What is the most recent GCC version you tried?

**Comment by plbossart** [comment]
This '-fanalyzer' option was introduced with GCC 10

https://gcc.gnu.org/onlinedocs/gcc-10.1.0/gcc/Static-Analyzer-Options.html

I have GCC11 with Ubuntu 22.10

File: sound/soc/intel/atom/sst/sst_stream.c [hunk 1]
```diff
@@ -217,7 +217,7 @@ int sst_send_byte_stream_mrfld(struct intel_sst_drv *sst_drv_ctx,
 		 * copy the reply and send back
 		 * we need to update only sz and payload
 		 */
-		if (bytes->block) {
+		if (bytes->block && block) {
```
Comment (paulstelian97) [inline sound/soc/intel/atom/sst/sst_stream.c]: As a comment on how static analyzers work (GCC or others), they _cannot_ reliably track null checks on anything that isn't a local variable, because of the chance of concurrent modification (whether from a signal, interrupt in kernel, or from a different core). From this point of view it's not a false positive. It's not something that a new version of the analyzer will fix.

The correct fix for such issues is to grab a copy of the pointer locally, and do the NULL check and usage on said local copy. If you know for a different reason that the pointer in the struct cannot be concurrently modified you can probably ignore the report as well.

This _specific_ fix isn't exactly to my liking, because it's bypassing the checker without really fixing what it points out.

TL;DR: GCC is right in a vacuum here. It is unable to guarantee the original code is correct because it cannot guarantee bytes->block doesn't change between check and use.

Comment (plbossart) [reply sound/soc/intel/atom/sst/sst_stream.c]: good feedback @paulstelian97, we can definitively add a local variable and make the code simpler.

Comment (paulstelian97) [reply sound/soc/intel/atom/sst/sst_stream.c]: As background on how I came up with this conclusion, I remember from a completely different language and personal projects (Kotlin's nullable pointers behave exactly as I explain, the compiler does not consider null pointer checks on non-local variables sufficient).

Comment (plbossart) [reply sound/soc/intel/atom/sst/sst_stream.c]: fixed with local variable as suggested

**Comment by marc-hb** [comment]
Debian/Ubuntu support multiple gcc versions in a single distribution release BUT only one default version. For Ubuntu 20.04, the default gcc version was gcc 9.3: https://packages.ubuntu.com/focal/devel/

I see you've updated to `ubuntu-latest` which currently points at ubuntu 22.04 which fixes the failure. I would not point at `ubuntu-latest` because it may change at a very inconvenient moment, I would go for `ubuntu-22.04` instead. Github has historically provided only LTS runners (used by 95% of Ubuntu users according to Ubuntu stats) so manually upgrading every 2 years seems like a much smaller and less risky inconvenience. No big deal, up to you!


https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners
https://ubuntu.com/about/release-cycle

**Comment by plbossart** [comment]
> I would not point at `ubuntu-latest` because it may change at a very inconvenient moment, I would go for `ubuntu-22.04` instead.

Yeah, that's what I had initially and then I changed my position. How would we know that something is broken with ubuntu-latest if we don't test it? In other words, let's use GitHub Actions to test GitHub actions. If things break we can always go back to the last version known to work.

**Comment by marc-hb** [comment]
>  How would we know that something is broken with ubuntu-latest if we don't test it?

I've never seen this being a real issue in any project. Linux distributions are very good with backwards compatibility, especially LTS releases. Now Ubuntu bugs and gcc bugs do exist of course (examples https://github.com/thesofproject/sof-test/discussions, https://github.com/thesofproject/sof/issues/7114,...) but I've never seen any affect something as common as compiling the kernel.

The converse is much more common: starting to use some brand new features without realizing it. That's why it's good for CI to always lag some (reasonable) distance behind developers. CI effectively defines the minimum requirements.

Moreever CI never stays stuck on old releases for a very long time because some new fancy feature like `-fanalyzer` is eventually found missing which forces an upgrade: exactly what is happening here right now.

Just sharing my experience, again no big deal.

File: .github/workflows/buildtest.yml [hunk 2]
```diff
@@ -39,9 +39,30 @@ jobs:
           make -k sound/soc/amd/ C=2
           make -k drivers/soundwire/ C=2
 
+  gcc-fanalyzer:
+    name: "GCC static analysis"
```
Comment (marc-hb) [inline .github/workflows/buildtest.yml]:
```suggestion
    name: "GCC static -fanalyzer"
```

File: include/linux/firmware/imx/dsp.h [hunk 3]
```diff
@@ -37,17 +37,11 @@ struct imx_dsp_ipc {
...
-	if (!ipc)
-		return;
-
 	ipc->private_data = data;
 }
 
 static inline void *imx_dsp_get_data(struct imx_dsp_ipc *ipc)
 {
-	if (!ipc)
-		return NULL;
```
Comment (RanderWang) [inline include/linux/firmware/imx/dsp.h:49]: this change is not aligned with most of changes in this PR. Can we check ipc before it is used ?

Comment (plbossart) [reply include/linux/firmware/imx/dsp.h:49]: I find this quite different, the other cases are more complicated and will have multiple cases and tests before returning NULL. Here it's just a straightforward test of something being NULL.

But that would be for @dbaluta @paulstelian97 to comment, I don't have any specific opinion on this code.

**Comment by plbossart** [comment]
No reply from @crojewsk-intel, dropped AVS patches and dropped AVS from the checks.

**Comment by crojewsk-intel** [comment]
Sorry for the late reply. My two cents (example based on bdw_rt286):

```
struct snd_soc_dai *codec_dai = snd_soc_card_get_codec_dai(card, "rt286-aif1");

if (!codec_dai)
	return 0;
```

in my opinion is an invalid scenario. Given the DAI LINKs declaration that is part of the machine board driver, such situation may never occur - if e.g.: `rt286-aif1` is missing, sound card won't even be enumerated.

Another thing to consider is correctness against PM expectations. That is, on suspend we should not prevent the system-wide suspend so 0 shall be returned regardless but on resume that's not the case and errors should be propagated. Of course, there's another level of complication as results `snd_soc_card_suspend_pre()` and `snd_soc_card_resume_post()` are completely ignored by the framework. Oh well..

**Review by RanderWang — DISMISSED** [review]

File: sound/soc/sof/intel/hda-dai.c [hunk 4]
```diff
@@ -341,6 +356,12 @@ static int non_hda_dai_hw_params(struct snd_pcm_substream *substream,
 	int stream_id;
 	int ret;
 
+	ops = hda_dai_get_ops(substream, cpu_dai);
```
Comment (ranj063) [inline sound/soc/sof/intel/hda-dai.c]: I think this is the only place this check makes sense in this patch. And the reason is that we'd try get the ops during hw_params nd if it is NULL, we'd never really ever get to trigger or link_dma_hw_params or link_dma_cleanup

File: sound/soc/sof/topology.c [hunk 5]
```diff
@@ -1117,10 +1117,11 @@ static void sof_disconnect_dai_widget(struct snd_soc_component *scomp,
 {
 	struct snd_soc_card *card = scomp->card;
 	struct snd_soc_pcm_runtime *rtd;
+	const char *sname = w->sname;
```
Comment (ranj063) [inline sound/soc/sof/topology.c:1120]: @plbossart how does this fix the warning about w being NULL?

File: sound/soc/intel/atom/sst/sst_stream.c [hunk 6]
```diff
@@ -174,10 +174,11 @@ int sst_send_byte_stream_mrfld(struct intel_sst_drv *sst_drv_ctx,
 	u32 length;
 	int pvt_id, ret = 0;
 	struct sst_block *block = NULL;
+	u8 bytes_block = bytes->block;
```
Comment (ranj063) [inline sound/soc/intel/atom/sst/sst_stream.c:176]: if this is a false positive, can we ignore it?

Comment (marc-hb) [reply sound/soc/intel/atom/sst/sst_stream.c:176]: Isn't a local variable clearer and better anyway? Faster and less concurrency.

File: sound/soc/sof/intel/hda-dai.c (hunk 4 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda-dai.c]: yes, but static analysis cannot know in which order the functions are called.

That's clearly explained in the commit message.

File: sound/soc/sof/topology.c (hunk 5 above)
Comment (plbossart) [reply sound/soc/sof/topology.c:1120]: w->sname could be modified by something else. a local variable cannot. See commit message and explanations from @paulstelian97

File: sound/soc/intel/atom/sst/sst_stream.c (hunk 6 above)
Comment (plbossart) [reply sound/soc/intel/atom/sst/sst_stream.c:176]: it's not a false positive @ranj063. Static analysis has to assume that bytes->block can be modified. Moving to a local variable removes this degree of freedom.

**Review by RanderWang — APPROVED** [review]

**Comment by plbossart** [comment]
> ```
> struct snd_soc_dai *codec_dai = snd_soc_card_get_codec_dai(card, "rt286-aif1");
> 
> if (!codec_dai)
> 	return 0;
> ```
> 
> in my opinion is an invalid scenario. Given the DAI LINKs declaration that is part of the machine board driver, such situation may never occur - if e.g.: `rt286-aif1` is missing, sound card won't even be enumerated.

Static analysis can't know that. It just looks at the result of snd_soc_card_get_codec_dai() and tells you there is one possible case of NULL dereference.

**Comment by plbossart** [comment]
@dbaluta @paulstelian97  can I ask you to look at the IMX part?

@yaochunhung @kuanhsuncheng can I ask you to look at the Mediatek part (same concept as IMX)?

**Comment by plbossart** [comment]
Unless I see a sustained objection, I am going to merge this series by the end of the week. 

I don't see any reason why we should not use a tool because of lack of comments/approvals. This PR has been around for 2 months, time to move forward.

@ranj063 @ujfalusi @bardliao  @dbaluta @paulstelian97 @yaochunhung @kuanhsuncheng @bhiregoudar FYI.

**Review by vijendarmukunda — APPROVED** [review]
LGTM

**Review by dbaluta — APPROVED** [review]

**Comment by dbaluta** [comment]
Looks good to me.

**Review by yaochunhung — APPROVED** [review]
LGTM. Thanks.

**Comment by marc-hb** [comment]
In https://github.com/thesofproject/sof/issues/7562 we found that there is apparently no way to manage false positives.
